packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs

```diff
@@ -146,169 +150,144 @@ impl PostgresDatabase {
+        database_type: &T,
         created_by: AccountId,
-    ) -> Result<(), InsertionError> {
-        sqlx::query_as::<_, (Uuid,)>(
+    ) -> Result<(), InsertionError>
+    where
+        T: DatabaseType + Serialize + Sync,
+    {
+        sqlx::query(&format!(
             r#"
-            INSERT INTO entity_types (
+            INSERT INTO {} (
                 version_id,
                 schema,
                 created_by
             ) 
-            VALUES ($1, $2, $3)
-            RETURNING version_id;
+            VALUES ($1, $2, $3);
             "#,
-        )
+            T::table()
+        ))
         .bind(version_id)
         .bind(
-            serde_json::to_value(entity_type)
-                .unwrap_or_else(|err| unreachable!("Could not serialize entity type: {err}")),
+            serde_json::to_value(database_type)
+                .unwrap_or_else(|err| unreachable!("Could not serialize data type: {err}")),
         )
         .bind(created_by)
-        .fetch_one(&self.pool)
+        .execute(&self.pool)
         .await
         .report()
-        .change_context(InsertionError)
-        .attach_lazy(|| entity_type.clone())?;
+        .change_context(InsertionError)?;
 
         Ok(())
     }
-}
-
-#[async_trait]
-impl Datastore for PostgresDatabase {
-    async fn create_data_type(
-        &mut self,
-        data_type: DataType,
-        created_by: AccountId,
-    ) -> Result<Qualified<DataType>, InsertionError> {
-        let uri = data_type.id();
-
-        if self
-            .contains_base_uri(uri.base_uri())
-            .await
-            .change_context(InsertionError)
-            .attach_lazy(|| data_type.clone())?
-        {
-            return Err(Report::new(InsertionError)
-                .attach_printable(BaseUriAlreadyExists)
-                .attach_printable(uri.base_uri().clone())
-                .attach(data_type.clone()));
-        }
-
-        self.insert_base_uri(uri.base_uri())
-            .await
-            .attach_lazy(|| data_type.clone())?;
-
-        let version_id = self
-            .insert_uri(uri)
-            .await
-            .attach_lazy(|| data_type.clone())?;
-
-        self.insert_data_type(version_id, &data_type, created_by)
-            .await
-            .attach_lazy(|| data_type.clone())?;
-
-        Ok(Qualified::new(version_id, data_type, created_by))
-    }
 
-    async fn get_data_type(
+    /// Returns the specified [`DatabaseType`].
+    ///
+    /// # Errors
+    ///
+    /// - If the specified [`VersionId`] does not already exist.
+    // TODO: We can't distinguish between an DB error and a non-existing version currently
+    async fn get_by_version<T: DatabaseType + DeserializeOwned>(
         &self,
         version_id: VersionId,
-    ) -> Result<Qualified<DataType>, QueryError> {
-        let (data_type, created_by) = sqlx::query_as(
+    ) -> Result<Qualified<T>, QueryError> {
+        let (data_type, created_by) = sqlx::query_as(&format!(
```

Comment:
Again, adding reasoning behind the `query safety` of interpolating the SQL would be nice here :)